cpp/references/data_structures/list_digraph.cpp

```diff
@@ -1 +1,288 @@
+}
+
+Array ListDigraph::add_vertices(Array p_data) {
+	Array ret = Array();
+	for (int i = 0; i < p_data.size(); i++) {
+		lemon::ListDigraph::Node vertex = _graph.addNode();
+		Vertex *instance = Vertex::_new(vertex, this);
+		(*_vertex_map)[vertex] = instance;
+		(*_vertex_data)[vertex] = p_data[i];
+		ret.append(instance);
+	}
+	return ret;
+}
+
+Arc *ListDigraph::add_arc(const Vertex *p_source, const Vertex *p_target) {
+	lemon::ListDigraph::Arc arc = _graph.addArc(p_source->get_vertex(), p_target->get_vertex());
+	Arc *instance = Arc::_new(arc, this);
+	(*_arc_map)[arc] = instance;
+	return instance;
+}
+
+void ListDigraph::erase_vertex(const Vertex *p_vertex) {
+	_graph.erase(p_vertex->get_vertex());
+}
+
+void ListDigraph::erase_arc(const Arc *p_arc) {
+	_graph.erase(p_arc->get_arc());
+}
+
+void ListDigraph::contract(const Vertex *p_vertex1, const Vertex *p_vertex2, bool r) {
+	_graph.contract(p_vertex1->get_vertex(), p_vertex2->get_vertex(), r);
+}
+
+void ListDigraph::clear() {
+	_graph.clear();
+}
+
+void ListDigraph::reserve_vertex (const int n) {
+	_graph.reserveNode(n);
+}
+
+void ListDigraph::reserve_arc (const int m) {
+	_graph.reserveArc(m);
+}
+
+Vertex *ListDigraph::get_vertex(const int p_id) {
+	lemon::ListDigraph::Node vertex = _graph.nodeFromId(p_id);
+	if (Ref<Vertex>((*_vertex_map)[vertex]).is_valid()) {
+		return Object::cast_to<Vertex>((*_vertex_map)[vertex]);
+	}
+	Vertex *instance = Vertex::_new(vertex, this);
+	(*_vertex_map)[vertex] = instance;
+	return instance;
+}
+
+Arc *ListDigraph::get_arc(const int p_id) {
+	lemon::ListDigraph::Arc arc = _graph.arcFromId(p_id);
+	if (Ref<Arc>((*_arc_map)[arc]).is_valid()) {
+		return Object::cast_to<Arc>((*_arc_map)[arc]);
+	}
+	Arc *instance = Arc::_new(arc, this);
+	(*_arc_map)[arc] = instance;
+	return instance;
+}
+
+Array ListDigraph::get_vertices() {
+	Array ret = Array();
+	for (lemon::ListDigraph::NodeIt it(_graph); it != lemon::INVALID; ++it) {
+		if (Ref<Vertex>((*_vertex_map)[it]).is_valid()) {
+			ret.append(Object::cast_to<Vertex>((*_vertex_map)[it]));
+		}
+		else {
+			Vertex *instance = Vertex::_new(it, this);
+			(*_vertex_map)[it] = instance;
+			ret.append(instance);
+		}
+
+	}
+	return ret;
+}
+
+Array ListDigraph::get_arcs() {
+	Array ret = Array();
+	for (lemon::ListDigraph::ArcIt it(_graph); it != lemon::INVALID; ++it) {
+		if (Ref<Arc>((*_arc_map)[it]).is_valid()) {
+			ret.append(Object::cast_to<Arc>((*_arc_map)[it]));
+		}
+		else {
+			Arc *instance = Arc::_new(it, this);
+			(*_arc_map)[it] = instance;
+			ret.append(instance);
+		}
+	}
+	return ret;
+}
+
+void ListDigraph::_register_methods() {
+	register_method("_iter_init", &ListDigraph::_iter_init);
+	register_method("_iter_next", &ListDigraph::_iter_next);
+	register_method("_iter_get", &ListDigraph::_iter_get);
+	
+	register_method("add_vertex", &ListDigraph::add_vertex);
+	register_method("add_vertices", &ListDigraph::add_vertices);
+	register_method("add_arc", &ListDigraph::add_arc);
```

Comment:
We should also create an `add_arcs` method, just like for vertices. Sorry if that wasn't clear earlier.